docs/proposals/external_providers_grpc.md

```diff
@@ -0,0 +1,48 @@
+# GRPC based implementation of Cloud Providers - WIP
+
+## Goal:
+Currently the Cloud Providers' (CP) functionalities ( Create(), Delete(), List() ) are part of the Machine Controller Manager's (MCM)repository. Because of this, adding support for new CPs into MCM requires merging code into MCM which may not be required for core functionalities of MCM itself. Also, for various reasons it may not be feasible for all CPs to merge their code with MCM which is an Open Source project.
+
+Because of these reasons, it was decided that the CP's code will be moved out in separate repositories so that they can be maintained separately by the respective teams. Idea is to make MCM act as a GRPC server, and CPs as GRPC clients. The CP can register themselves with the MCM using a GRPC service exposed by the MCM. Details of this approach is discussed below.
+
+## How it works:
+MCM acts as GRPC server and listens on a pre-defined port 5000. It implements below GRPC services. Details of each of these services are mentioned in next section.
+* `Register()`
+* `GetMachineClass()`
+* `GetSecret()`
+
+## GRPC services exposed by MCM:
+
+### Register()
+`rpc Register(stream DriverSide) returns (stream MCMside) {}`
+
+The CP GRPC client calls this service to register itself with the MCM. The CP passes the `kind` and the `APIVersion` which it implements, and MCM maintains an internal map for all the registered clients. A GRPC stream is returned in response which is kept open througout the life of both the processes. MCM uses this stream to communicate with the client  for machine operations: `Create()`, `Delete()` or `List()`.
```

Comment:
> MCM maintains an internal map for all the registered clients

Is it possible to avoid keeping the state in-memory ?
* if MCM restarts it looses the state. Driver then needs to to figure out the restart incident and re-register. 

How about persisting the driver-name[provisioner in storageclasses] - in the MachineClass ?? Can you please add this as a possible way - we can then check the implementation details. 
For example in StorageClass:  https://kubernetes.io/docs/concepts/storage/storage-classes/#aws

Reply:
If MCM goes down, the connection/stream also breaks. Thus, driver anyway need to create a new connection in that case.

Reply:
I agree with @ggaurav10 and @amshuman-kr on this.